music21/key.py

```diff
@@ -996,6 +996,56 @@ def tonicPitchNameWithCase(self):
             tonic = tonic.lower()
         return tonic
 
+    def deriveByKeySignature(self, KeySig: Union[KeySignature, str]) -> List['Key']:
+        '''
+        Derive a fitting Key from the KeySigntature and the tonic of this Key.
+        Thew new key will have the same tonic as this key and the elternated pitches from
+        the key and the key signature. Conflicting alternating keys are overriden by the
+        key signature.
+
+        >>> k = key.Key('G', 'major')
+        >>> ks = key.KeySignature()
+        >>> ks.alteredPitches = ['Fn']
+        >>> k.deriveByKeySignature(ks)
+        [<music21.key.Key of G mixolydian>]
+
+        >>> k = key.Key('C', 'major')
+        >>> ks = key.KeySignature()
+        >>> ks.alteredPitches = ['F#']
+        >>> k.deriveByKeySignature(ks)
+        [<music21.key.Key of C lydian>, <music21.key.Key of C hypolocrian>]
+        '''
+
+        def derived_scales(cls: Type):
+            for subscale in cls.__subclasses__():
+                yield subscale
+                yield from derived_scales(subscale)
+
+        # override the pitches of this key with the pitches of the key scale
+        keyPitches = {p.step: p for p in self.pitches}
+        for alteredPitch in KeySig.alteredPitches:
+            keyPitches[alteredPitch.step] = alteredPitch
+
+        # we are looking from ConcreteScales derived Scale
+        scales = []
+        for name in derived_scales(scale.ConcreteScale):
+            # find closest-matching ConcreteScale based on the pitch collection
+            try:
+                # Derive a scale from the concrete scale using the keyPitches
+                sc = name(self.tonic).derive(other=keyPitches.values())
+                if sc.tonic == self.tonic:
+                    # create a key with the given tonic and  mode of the scale
+                    scales.append(Key(tonic=self.tonic, mode=sc.type))
+            except AttributeError:
+                # Something has failed within one of the scales
+                pass
```

Comment:
Thanks for finding the issue with intervalList! I have a separate patch for this in #785, so I suggest removing the workaround here.

Reply:
i did.